src/DoubleTheDonation/FormExtension/Actions/ConvertBlockToField.php

```diff
@@ -0,0 +1,79 @@
+<?php
+
+namespace GiveDoubleTheDonation\DoubleTheDonation\FormExtension\Actions;
+
+use Give\Donations\Models\Donation;
+use Give\Framework\Blocks\BlockModel;
+use Give\Framework\FieldsAPI\Contracts\Node;
+use Give\Framework\FieldsAPI\Exceptions\EmptyNameException;
+use GiveDoubleTheDonation\DoubleTheDonation\FormExtension\Field as DoubleTheDonationField;
+use GiveDoubleTheDonation\DoubleTheDonation\Helpers\DoubleTheDonationApi;
+
+/**
+ * @unreleased
+ */
+class ConvertBlockToField
+{
+    /**
+     * @unreleased
+     * @throws EmptyNameException
+     */
+    public function __invoke(?Node $node, BlockModel $block, int $blockIndex, int $formId): ?Node
+    {
+        if ( ! give(DoubleTheDonationApi::class)->isKeyValid()) {
+            return null;
+        }
+
+        return DoubleTheDonationField::make('dtd')
+            ->tap(function (DoubleTheDonationField $field) use ($block) {
+                $this->setFieldAttributes($field, $block);
+                $this->handleFieldScope($field);
+            });
+    }
+
+    /**
+     * Set field attributes from block attributes
+     *
+     * @unreleased
+     */
+    private function setFieldAttributes(DoubleTheDonationField $field, BlockModel $block): void
+    {
+        // set props from block attributes
+        $field->setLabel($block->getAttribute('label'));
+    }
+
+    /**
+     * Handle field scope
+     *
+     * @unreleased
+     */
+    private function handleFieldScope(DoubleTheDonationField $field): void
+    {
+        $field->scope(function (DoubleTheDonationField $field, $data, Donation $donation) {
```

Comment:
I like to turn the logic inside these callback scopes into actions that we can then write a unit test for so we're confident in the persistence of the field.

Reply:
Currently, you can't pass an action (or I'm not familiar with the way of doing that). I think we will have to update the FieldsAPI to support that. This might be an interesting idea to tackle on Fridays